xrpl/models/utils.py

```diff
@@ -68,6 +110,151 @@ def validate_credential_ids(credential_list: Optional[List[str]]) -> Dict[str, s
+
+    validation_messages: List[str] = []
+
+    if bool(HEX_REGEX.fullmatch(input_hex)) is False:
+        validation_messages.append("MPTokenMetadata must be in hex format.")
+        return validation_messages
+
+    if len(input_hex) == 0 or len(input_hex) > MAX_MPTOKEN_METADATA_LENGTH:
+        validation_messages.append(
+            (
+                "MPTokenMetadata must be non-empty and max "
+                f"{int(MAX_MPTOKEN_METADATA_LENGTH / 2)} bytes."
+            )
+        )
+        return validation_messages
+
+    try:
+        decoded_str = hex_to_str(input_hex)
+        json_metadata = json.loads(decoded_str)
+    except (json.JSONDecodeError, UnicodeDecodeError) as e:
+        validation_messages.append(
+            f"MPTokenMetadata is not properly formatted as JSON - {str(e)}"
+        )
+        return validation_messages
+
+    if not isinstance(json_metadata, Dict):
+        validation_messages.append(
+            "MPTokenMetadata is not properly formatted as per XLS-89d."
+        )
+        return validation_messages
+
+    # Structure validation
+
+    field_count = len(json_metadata.keys())
+    if field_count > MAX_MPT_META_TOP_LEVEL_FIELD_COUNT:
+        validation_messages.append(
+            (
+                "MPTokenMetadata must not contain more than "
+                f"{MAX_MPT_META_TOP_LEVEL_FIELD_COUNT} top-level fields "
+                f"(found {field_count})."
+            )
+        )
+        return validation_messages
+
+    for field in MPT_META_REQUIRED_FIELDS:
+        if not isinstance(json_metadata.get(field), str):
+            validation_messages.append(f"{field} is required and must be string.")
+
+    if len(validation_messages) > 0:
+        return validation_messages
+
+    if "desc" in json_metadata and not isinstance(json_metadata["desc"], str):
+        validation_messages.append("desc must be a string.")
+        return validation_messages
+
+    if "asset_subclass" in json_metadata and not isinstance(
+        json_metadata["asset_subclass"], str
+    ):
+        validation_messages.append("asset_subclass must be a string.")
+        return validation_messages
+
+    additional_info = json_metadata.get("additional_info")
+    if additional_info is not None and not (
+        isinstance(additional_info, str) or isinstance(additional_info, dict)
+    ):
+        validation_messages.append("additional_info must be a string or JSON object.")
+        return validation_messages
+
+    if "urls" in json_metadata:
+        urls = json_metadata["urls"]
+        if not isinstance(urls, list):
+            validation_messages.append("urls must be an array as per XLS-89d.")
+            return validation_messages
+
+        if not all(_is_valid_mpt_url_structure(u) for u in urls):
+            validation_messages.append(
+                "One or more urls are not structured per XLS-89d."
+            )
+            return validation_messages
```

Comment:
nit: In the last few if-conditions, you are immediately executing `return` statement after capturing the warning message. It is not useful to append to the `validation_messages` list. The control returns at most one message per execution of the method.